### PR TITLE
Classify on pi

### PR DESCRIPTION
**Pull request by gferraro**

### CODE REVIEW

File: ml_tools/trackdatabase.py [hunk 1]
```diff
@@ -7,46 +7,39 @@
...
-
-HDF5_LOCK = Lock()
-
 
 class HDF5Manager:
     """ Class to handle locking of HDF5 files. """
 
+    LOCK_FILE = "/var/lock/classifier-hdf5.lock"
+
     def __init__(self, db, mode="r"):
```
Comment (mjs) [inline ml_tools/trackdatabase.py]: Let's rename `db`  to `filename` to make it clear it's not a database object

File: ml_tools/trackdatabase.py [hunk 2]
```diff
@@ -7,46 +7,39 @@
...
-#
-# pool = multiprocessing.Pool(self.workers_threads, initializer=init_workers, initargs=(shared_lock,))
-
-HDF5_LOCK = Lock()
-
 
 class HDF5Manager:
     """ Class to handle locking of HDF5 files. """
 
+    LOCK_FILE = "/var/lock/classifier-hdf5.lock"
```
Comment (mjs) [inline ml_tools/trackdatabase.py]: A single lock file location could cause problems when the code is being run more than once a single machine. Could the lock file live next to the DB file?  Just append `.lock` to the DB filename.

File: ml_tools/trackdatabase.py [hunk 3]
```diff
@@ -93,18 +86,35 @@ def create_clip(self, clip, overwrite=True):
 
             if clip is not None:
                 group_attrs = group.attrs
-                group_attrs.update(clip.background_stats)
+
+                # group_attrs.update(clip.stats)
```
Comment (mjs) [inline ml_tools/trackdatabase.py]: is this supposed to be there? delete if not required.

File: piclassifier/cptvtest.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import time
+import os
+from cptv import CPTVReader
+import sys
+import numpy as np
+import h5py
+
+SOCKET_NAME = "/var/run/lepton-frames"
+test_cptv = "test.cptv"
+test_h5py = "/home/zaza/Cacophony/classifier-pipeline/pithermal65.h5py"
```
Comment (mjs) [inline piclassifier/cptvtest.py]: These appear to be unused (and we don't want paths that only make sense on your machine anyway)

File: piclassifier/cptvtest.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/python3
```
Comment (mjs) [inline piclassifier/cptvtest.py]: Add some docs at the top to explain what this does?

File: piclassifier/locationconfig.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+CONFIG_FILENAME = "location.yaml"
+CONFIG_DIRS = [Path(__file__).parent.parent, Path("/etc/cacophony")]
+
+
+@attr.s
+class LocationConfig:
+
+    DEFAULT_LAT = -43.5321
+    DEFAULT_LONG = 172.6362
```
Comment (mjs) [inline piclassifier/locationconfig.py]: Don't set a default location if it's missing!

The only time it's ok to use a default location is for sunset/sunrise calculations - not for recordings.

File: piclassifier/telemetry.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+def get_uint64(raw, offset):
+    return (
+        raw[offset + 1]
+        | (raw[offset] << 8)
+        | (raw[offset + 3] << 16)
+        | (raw[offset + 2] << 24)
+        | (raw[offset + 5] << 32)
+        | (raw[offset + 4] << 40)
+        | (raw[offset + 7] << 48)
+        | (raw[offset + 6] << 56)
```
Comment (mjs) [inline piclassifier/telemetry.py]: It might be faster to use the struct package for these conversions

File: requirements.txt [hunk 8]
```diff
@@ -1,7 +1,7 @@
 tensorflow~=1.12
 matplotlib~=3.0
 pytz
-cptv~=0.3.0
+cptv==1.0.0
```
Comment (mjs) [inline requirements.txt]: this is up to 1.1.0 now

File: track/framebuffer.py [hunk 9]
```diff
@@ -17,116 +17,166 @@
...
+        :param opt_flow: An optical flow algorithm
         """
+        height, width = self.thermal.shape
+        flow = np.zeros([height, width, 2], dtype=np.float32)
+        threshold = np.median(self.thermal) + flow_threshold
+        scaled_thermal = np.uint8(np.clip(self.thermal - threshold, 0, 255))
+        if prev_frame is not None:
+            # for some reason openCV spins up lots of threads for this which really slows things down, so we
+            # cap the threads to 2
+            cv2.setNumThreads(2)
```
Comment (mjs) [inline track/framebuffer.py]: I wonder if this should be `1` on a pi or are we not using optical flow at all?

File: track/trackextractor.py [hunk 10]
```diff
@@ -223,7 +218,7 @@ def calculate_preview(self, frame_list):
             logging.error("Video consists entirely of preview")
             number_frames = len(frame_list)
         frames = np.int32(frame_list[0:number_frames])
-        background = np.average(frames, axis=0)
+        background = np.min(frames, axis=0)
```
Comment (mjs) [inline track/trackextractor.py]: to discuss

File: track/trackextractor.py [hunk 11]
```diff
@@ -676,7 +662,23 @@ def analyse_background(self, frames):
 
         return background, background_stats
 
+    def set_optical_flow_function(self):
+        if not self.opt_flow:
```
Comment (mjs) [inline track/trackextractor.py]: How about:

```python
if self.opt_flow:
    return
```

File: track/trackextractor.py [hunk 12]
```diff
@@ -676,7 +662,23 @@ def analyse_background(self, frames):
...
+            self.opt_flow = cv2.createOptFlow_DualTVL1()
+            self.opt_flow.setUseInitialFlow(True)
+            if not self.config.high_quality_optical_flow:
```
Comment (mjs) [inline track/trackextractor.py]: Same here. Flip the condition and return early.

File: classify/main.py [hunk 13]
```diff
@@ -3,14 +3,17 @@
 import logging
 from datetime import datetime
 
+from .clipclassifier import ClipClassifier
 from ml_tools.logs import init_logging
 from ml_tools import tools
 from ml_tools.config import Config
 from ml_tools.previewer import Previewer
-from .clipclassifier import ClipClassifier
+import absl.logging
```
[the remark this answers is not in the document]
Comment (gferraro) [reply classify/main.py]: a bug with tensor flow preventing logging

File: piclassifier/telemetry.py (hunk 7 above)
Comment (gferraro) [reply piclassifier/telemetry.py]: i think should put this in python-cptv

File: track/framebuffer.py (hunk 9 above)
Comment (gferraro) [reply track/framebuffer.py]: not using it at all

File: piclassifier/motiondetector.py [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+        self.location = Location()
+        self.location.latitude = location_config.latitude
+        self.location.longitude = location_config.longitude
+
+        self.location.altitude = location_config.altitude
+        self.location.timezone = tools.get_timezone_str(
+            location_config.latitude, location_config.longitude
+        )
+
+    def set_sunrise_sunet(self):
```
Comment (mjs) [inline piclassifier/motiondetector.py]: get_sunrise_sunset

File: piclassifier/motiondetector.py [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+                            self.background_weight,
+                        )
+                    )
+                    temp_changed = True
+                    self.background_weight = (
+                        MotionDetector.BACKGROUND_WEIGHTING_PER_FRAME
+                    )
+
+            if (
+                temp_changed is False
```
Comment (mjs) [inline piclassifier/motiondetector.py]: `not temp_changed`

**Comment by gferraro** [comment]
Ran locally with cacophony-processing and cacophony-api and cacophony-browse.
Also compared the metadata and seems to be all good

**Pull request merged**